Add taskflow debug observer

Levi-Armstrong opened the pull request:
This allow you enable debug with in a ProcessManager which will attache and observer which will output additional information as ti executes the taskfflow.

mpowelson commented on tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/debug_observer.h:
```diff
@@ -0,0 +1,54 @@
+ * @file debug_observer.h
+ * @brief Debug Taskflow Observer
+ *
+ * @author Matthew Powelson
+ * @date July 15. 2020
```
Needs updating

Levi-Armstrong replied:
I copied this from the you created so kept the same date.

mpowelson replied:
Oh ok. Got it.

mpowelson commented on tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/debug_observer.h:
```diff
@@ -0,0 +1,54 @@
+
+namespace tesseract_planning
+{
+class DebugObserver : public tf::ObserverInterface
```
Doxygen

Levi-Armstrong replied:
Done

mpowelson commented on tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_generators/continuous_contact_check_process_generator.h:
```diff
@@ -68,7 +68,7 @@ class ContinuousContactCheckProcessGenerator : public ProcessGenerator
   std::string name_;
 
-  double longest_valid_segment_length_{ 0.1 };
+  double longest_valid_segment_length_{ 0.005 };
```
This seems pretty small.

Levi-Armstrong replied:
Yea, I will try 0.05 instead. The challenge I keep running into with continuous collision checking is that if you have two kinematic links that are always within say 2cm and are not disabled in ACM. Then if you move more than 2cm in a single time step it will report collision because the end state of the first shape overlaps with the start state of the second shape. This is why I have switched to using Discrete Continuous collision checking for motion planning and I am getting much better performance then I run the post check a fine resolution.

Levi-Armstrong replied:
Also I did not mean to include this in this PR. I will remove it and create a different PR to discuss further.

mpowelson replied:
I was actually running into the same thing. Do you think DISCRETE_CONTINUOUS or SINGLE_TIMESTEP should be the default? It is also more difficult to debug in this framework since the trajopt plotting has to sort of be hacked in there.

mpowelson replied:
Is `composite_profile->collision_constraint_config.type = trajopt::CollisionEvaluatorType::DISCRETE_CONTINUOUS;` all you have to do to set that? I'm still getting blue arrows

Levi-Armstrong replied:
> I was actually running into the same thing. Do you think DISCRETE_CONTINUOUS or SINGLE_TIMESTEP should be the default? It is also more difficult to debug in this framework since the trajopt plotting has to sort of be hacked in there.

I do believe we should set the default to Discrete Continuous. I think because of the previous issue it makes motion planning difficult to debug. I do think the post check is still good to use continuous but you need to do this at a fine resolution to no run into issues.

Levi-Armstrong replied:
> Is `composite_profile->collision_constraint_config.type = trajopt::CollisionEvaluatorType::DISCRETE_CONTINUOUS;` all you have to do to set that? I'm still getting blue arrows

You should still get the blue arrows with Discrete continuous. When interpolating between two states it stores the contact results with the deepest penetration and then shows the contact point on the start and end state like what is done with Continuous Collision checking.

Levi-Armstrong replied:
I reset it for this PR and will open up a new one changing these settings.

Levi-Armstrong replied:
I have create PR #392 to further discuss with proposed changes.

mpowelson commented on tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_manager.h:
```diff
@@ -73,6 +73,25 @@ class ProcessManager
+   *
+   * @param enabled
+   */
+  virtual void setDebug(bool enabled) = 0;
+
+  /**
+   * @brief Enable/Disable profiling for the process manager
+   *
+   * This will add an observer which will profile task as it is executed. It will save a file in the temp directory
+   * which can be loaded into the taskflow profile webpage.
+   *
+   * @param enabled
+   */
+  virtual void setProfile(bool enabled) = 0;
```
Maybe enableDebug and enableProfiling? SetProfile sounds like the motion planning profiles

Levi-Armstrong replied:
I like it.

mpowelson reviewed (approved):
I got excited for a minute. Can't wait to see the actual debug and profiling observers.

Levi-Armstrong commented:
> I got excited for a minute. Can't wait to see the actual debug and profiling observers.

I just added the profile observer take a look and let me know if anything should be changed. I switched to adding a string timestamp to the file names.

Levi-Armstrong commented:
It will save a file in the tmp directory and you can copy the contents into this [tool](https://taskflow.github.io/tfprof/).

mpowelson commented:
Did you push it? I don't see it

Levi-Armstrong commented:
Sorry I forced pushed. The observer is already available in Taskflow I just needed to create it and add it to the executor.

Levi-Armstrong commented on tesseract/tesseract_planning/tesseract_process_managers/src/process_managers/raster_dt_process_manager.cpp:
```diff
@@ -201,6 +207,14 @@ bool RasterDTProcessManager::execute()
 {
   success_ = true;
 
+  DebugObserver::Ptr debug_observer;
+  std::shared_ptr<tf::TFProfObserver> profile_observer;
+  if (debug_)
+    debug_observer = executor_.make_observer<DebugObserver>("RasterDTProcessManagerObserver");
+
+  if (profile_)
+    profile_observer = executor_.make_observer<tf::TFProfObserver>();
```
Updated section along with saving it to a file below.

Levi-Armstrong replied:
I am also updating in the Planning Request on the ROS side

mpowelson replied:
Oh ok. I didn't realize it was in Taskflow. I'll have to try it sometime

Levi-Armstrong replied:
Yea neither did I until recently. I ask a question looking for a way to trace the path taken through the graph and they pointed me to this.
